### PR TITLE
Add support for client-side QUIC 0-RTT

### LINKED ISSUE #16234: regression test QUIC crash caused by the import
See discussion on #16100 (h/t @moderation for discovery)

### PR DESCRIPTION
**Pull request by DavidSchinazi**
This PR introduces EnvoyQuicSessionCache, which is responsible
for saving QUIC+TLS session tickets which are required for 0-RTT.

I've confirmed that the new integration test from this PR would
reproduce the 0-RTT crash that we saw previouslybut had since
been fixed.

Risk Level: Low
Testing: Local
Docs Changes: None
Release Notes: None
Platform Specific Features: None

Closes #16234

### CODE REVIEW

File: source/common/quic/envoy_quic_session_cache.h
Comment (rojkov) [inline source/common/quic/envoy_quic_session_cache.h:57]: A drive-by comment: wouldn't it be better to use a hash map hear since the code doesn't seem to rely on the order of elements? Also `absl::btree_map` is used usually for ordered maps.

Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h:57]: A hash map requires a hash function to be defined, which unfortunately isn't the case here.
absl::btree_map requires the copy operator to be defined, which also isn't the case here.
Given that this only used at connection creation time, the performance impact is negligible compared to related asymmetric cryptographic operations so I'd prefer to keep std::map for now.

File: source/common/quic/envoy_quic_session_cache.h
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.h:42]: Why size of 2?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.cc:130]: It's a bit confusing about the life time of application_state. Can you comment that it's now owned?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.cc:138]: nits: consider wrapping these into Entry constructor.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.cc:108]: Is it safe to erase the entry while iterating on the map?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.cc:7]: Envoy doesn't use kCamelCaseConstantName convention for const. MAX_SESSION_CACHE_SIZE instead?
Do this needs to be configuration? @alyssawilk

**Review by danzh2010 — COMMENTED** [review]
Nice, this should also address #16146.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:10]: There is no `system_time_` member. I think you might be confusing it with `time_source_` which is different - we want to avoid computing the time from the time source too often.

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h:42]: Because that's sufficient in practice, it's the value Chrome uses.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:108]: Yes. It invalidates the iterator but we don't use it afterwards

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:7]: I went with `MaxSessionCacheEntries` which is preferred by the [Envoy style guide](https://github.com/envoyproxy/envoy/blob/main/STYLE.md) because it doesn't conflict with macros

Comment (alyssawilk) [reply source/common/quic/envoy_quic_session_cache.cc:7]: we'll want config eventually but that can be a TODO (either for David, against QUIC upstream, I vote for the former =P) with a comment that the hard-coded value is arbitrary

Also @RyanTheOptimist look, another bounded cache!  now that I look they're everywhere =P

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:138]: Given that this is a struct which needs a default constructor, I think this is overall simpler?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:130]: The input parameter owned. The fact that it's a const pointer makes the ownership model clear to me. Am I missing something?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc]: I can file an issue to make this configurable once this lands. Though I don't think it's a priority

Comment (alyssawilk) [reply source/common/quic/envoy_quic_session_cache.cc]: I'm thinking more that having configurable 0-rtt config and integration tests is worthwhile for envoy mobile :-)

Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc]: What's a configurable 0-RTT config?
This PR adds an integration test.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_session_cache.cc:20]: How costly are there two functions? Worth latching these two in Entry?

Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:20]: They're very cheap, just dereferencing the session pointer.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:130]: > sorry, typo "not" for "now".

I don't think we need to comment that a const pointer is not owned, as that's very common usage.

> If we create ApplicationState instance for each entry here, why does doApplicationStatesMatch() compare pointers?

doApplicationStatesMatch() dereferences the pointers and compares the values. It looks at pointers because they can be nullptr.

> And Lookup() also create ApplicationState for return value. Probably QUICHE somehow requires duplication this object.

I'm not sure what you're asking? QUICHE passes us a const pointer and we need to make a copy that we own.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:10]: Makes sense. I've renamed `system_time` to `now`.

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h:42]: Added a comment.

File: source/common/quic/envoy_quic_session_cache.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_session_cache.h:48]: Optionally
If after that we are still at the size limit, also remove ->
If all entries were valid but the cache is at its size limit, instead removes

File: test/common/quic/envoy_quic_session_cache_test.cc
Comment (alyssawilk) [inline test/common/quic/envoy_quic_session_cache_test.cc:100]: mind commenting why you can't just use simtime's monotonicTime too?

File: test/common/quic/envoy_quic_session_cache_test.cc
Comment (alyssawilk) [inline test/common/quic/envoy_quic_session_cache_test.cc:181]: I know they're not used in all tests, but would it make sense to make some of these member variables just to cut down on test set-up?

**Review by alyssawilk — COMMENTED** [review]
this looks fantastic - it'll be great to have 0-rtt working and even better to regression-proof that crash :-)

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:65]: Some clients may want to use different CCL_CTX'es. We could remove it but it isn't causing any issues. Either way that would be a QUICHE CL.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:9]: Done

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc]: Done

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:44]: What's QuicClientSessionCache?

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc]: Done

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:133]: O(1024) is negligible compared to the cryptographic operations required to create a new connection.

Added comments though

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h]: Done

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h:45]: Done

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h]: Done

File: test/common/quic/envoy_quic_session_cache_test.cc
Comment (DavidSchinazi) [reply test/common/quic/envoy_quic_session_cache_test.cc]: Done

File: test/common/quic/envoy_quic_session_cache_test.cc
Comment (DavidSchinazi) [reply test/common/quic/envoy_quic_session_cache_test.cc:186]: I moved params to a member but not session because the session ownership is transferred to the cache

File: source/common/quic/envoy_quic_session_cache.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_session_cache.cc:44]: https://chromium.googlesource.com/chromium/src/+/master/net/quic/quic_client_session_cache.cc :-)

File: source/common/quic/envoy_quic_session_cache.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_session_cache.cc:67]: yeah, sorry, quiche == upstream in this context.  Not something you'd do in situ but if it's not used anywhere might be worth adding our quiche clean up list.

File: source/common/quic/envoy_quic_session_cache.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_session_cache.h:33]: ditto | here and below

**Review by alyssawilk — COMMENTED** [review]
LG modulo one more | nit!

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:67]: SG

File: source/common/quic/envoy_quic_session_cache.h
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.h]: Done

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc]: Done

File: source/common/quic/envoy_quic_session_cache.cc
Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_session_cache.cc:32]: Yes, this is needed to avoid a dereference when they're both nullptr. I personally prefer this style of comparison because it is correct for all inputs. The fact that callers won't ever send the same non-null pointer as both parameters is an implementation details that doesn't need to be visible in this function.

File: source/common/quic/envoy_quic_session_cache.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_session_cache.cc:45]: Yep, but modulo "can we refactor away the time abstraction" seems reasonable to stick this one in quiche at some point.

**Review by alyssawilk — APPROVED** [review]
LGTM, modulo Dan's sign-off

File: test/common/quic/envoy_quic_session_cache_test.cc
Comment (danzh2010) [inline test/common/quic/envoy_quic_session_cache_test.cc:23]: nits: renaming these const according to Envoy naming convention. (some of them are only used once, plz consider inlining them)

Comment (DavidSchinazi) [reply test/common/quic/envoy_quic_session_cache_test.cc]: Renamed the constants to the Envoy naming convention. I kept the ones used only once because they make the code more readable.

**Review by danzh2010 — APPROVED** [review]
LGTM, thanks for adding this!

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
